test/dialing/resolver.spec.js

```diff
@@ -0,0 +1,188 @@
+  })
+
+  it('resolves a dnsaddr recursively', async () => {
+    const remoteId = remoteLibp2p.peerId.toB58String()
+    const dialAddr = multiaddr(`/dnsaddr/remote.libp2p.io/p2p/${remoteId}`)
+    const relayedAddrFetched = multiaddr(relayedAddr(remoteId))
+
+    // Transport spy
+    const transport = libp2p.transportManager._transports.get('Circuit')
+    sinon.spy(transport, 'dial')
+
+    // Resolver stub
+    const stub = sinon.stub(Resolver.prototype, 'resolveTxt')
+    let firstCall = false
+    stub.callsFake(() => {
+      if (!firstCall) {
+        firstCall = true
+        // Return an array of dnsaddr
+        return Promise.resolve(getDnsaddrStub(remoteId))
+      }
+      return Promise.resolve(getDnsRelayedAddrStub(remoteId))
+    })
+
+    // Dial with address resolve
+    const connection = await libp2p.dial(dialAddr)
+    expect(connection).to.exist()
+    expect(connection.remoteAddr.equals(relayedAddrFetched))
+
+    const dialArgs = transport.dial.firstCall.args
+    expect(dialArgs[0].equals(relayedAddrFetched)).to.eql(true)
+  })
+
+  // TODO: Temporary solution does not resolve dns4/dns6
+  // Resolver just returns the received multiaddrs
+  it('stops recursive resolve if finds dns4/dns6 and dials it', async () => {
+    const remoteId = remoteLibp2p.peerId.toB58String()
+    const dialAddr = multiaddr(`/dnsaddr/remote.libp2p.io/p2p/${remoteId}`)
+
+    // Stub resolver
+    const dnsMa = multiaddr(`/dns4/ams-1.remote.libp2p.io/tcp/443/wss/p2p/${remoteId}`)
+    const stubResolve = sinon.stub(Resolver.prototype, 'resolveTxt')
+    stubResolve.returns(Promise.resolve([
+      [`dnsaddr=${dnsMa}`]
+    ]))
+
+    // Stub transport
+    const transport = libp2p.transportManager._transports.get('WebSockets')
+    const stubTransport = sinon.stub(transport, 'dial')
+    stubTransport.callsFake((multiaddr) => {
+      expect(multiaddr.equals(dnsMa)).to.eql(true)
+    })
+
+    await libp2p.dial(dialAddr)
+  })
+
+  it('resolves a dnsaddr recursively not failing if one address fails to resolve', async () => {
+    const remoteId = remoteLibp2p.peerId.toB58String()
+    const dialAddr = multiaddr(`/dnsaddr/remote.libp2p.io/p2p/${remoteId}`)
+    const relayedAddrFetched = multiaddr(relayedAddr(remoteId))
+
+    // Transport spy
+    const transport = libp2p.transportManager._transports.get('Circuit')
+    sinon.spy(transport, 'dial')
+
+    // Resolver stub
+    let firstCall = false
+    let secondCall = false
+
+    const stub = sinon.stub(Resolver.prototype, 'resolveTxt')
+    stub.callsFake(() => {
+      if (!firstCall) {
+        firstCall = true
+        // Return an array of dnsaddr
+        return Promise.resolve(getDnsaddrStub(remoteId))
+      } else if (!secondCall) {
+        secondCall = true
+        // Address failed to resolve
+        return Promise.reject(new Error())
+      }
+      return Promise.resolve(getDnsRelayedAddrStub(remoteId))
+    })
```

Comment:
You should be able to use sinon to simplify this a bit, something like:
```suggestion
    const stub = sinon.stub(Resolver.prototype, 'resolveTxt')
    stub.onCall(0).callsFake(() => Promise.resolve(getDnsaddrStub(remoteId)))
    stub.onCall(1).callsFake(() => Promise.reject(new Error()))
    stub.callsFake(() => Promise.resolve(getDnsRelayedAddrStub(remoteId)))
```

Reply:
Nice, that works. I did not understand that I could use callsFake for the "other" calls, which was making things difficult 👍